### PR TITLE
Play things from queue

### PR DESCRIPTION
**Pull request by reidbaker**
This branch allows songs to be played from the playlist class and updates various ui components. Please checkout code and run on your device.

More importantly please check _small_ things like properly names variables, good comments for confusing code and formatting as well as _big_ things like architecture and the way information is passed.

### CODE REVIEW

File: fanClub/res/layout/content_frame.xml [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
-	    android:id="@+id/content_frame"
-	    android:layout_width="match_parent"
-	    android:layout_height="match_parent"/>   
-	
-	<FrameLayout
-        android:id="@+id/playbar"
-        android:layout_alignParentBottom="true"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"/>
+    <FrameLayout
```
Comment (bryansills) [inline fanClub/res/layout/content_frame.xml:4]: You messed up the formatting on this.

File: fanClub/src/com/lastcrusade/fanclub/CoreActivity.java [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 import android.os.Bundle;
 import android.support.v4.app.Fragment;
+import android.util.Log;
+import android.view.MotionEvent;
```
Comment (bryansills) [inline fanClub/src/com/lastcrusade/fanclub/CoreActivity.java]: looks like unnecessary imports

File: fanClub/src/com/lastcrusade/fanclub/util/MusicListAdapter.java [hunk 3]
```diff
@@ -67,7 +67,7 @@ public View getView(int position, View convertView, ViewGroup parent) {
 
         addButton.setTag(position);
 
-        userColor.setBackgroundColor(Color.parseColor(users.get(metadataList.get(position).getUsername())));
+        userColor.setBackgroundColor(0xFFFFFFFF);
```
Comment (dgreenhalgh) [inline fanClub/src/com/lastcrusade/fanclub/util/MusicListAdapter.java]: Could this be not hardcoded?

File: fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java [hunk 4]
```diff
@@ -34,23 +49,29 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
...
+
+        playPause.setOnClickListener(new OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                try {
+                    PlaylistService service = playlistServiceLocator.getService();
+                    if (service.isPlaying()) {
+                        service.pause();
+                        setPlayImage(playPause);
+                        Log.w(TAG, "pause called");
```
Comment (bryansills) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java]: should not be a warning

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java [hunk 5]
```diff
@@ -68,10 +64,19 @@ public PlaylistService getService() {
 
     private BroadcastRegistrar    registrar;
     private SingleFileAudioPlayer audioPlayer;
+    private Playlist queue;
```
Comment (dgreenhalgh) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: What's the reason for changing this variable name? Seems like @ejohnson44's evil doing

File: fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java [hunk 6]
```diff
@@ -68,4 +89,30 @@ public void onClick(View v) {
 
         return view;
     }
+
+    private void setPlayImage(final ImageButton playPause) {
```
Comment (bryansills) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java]: why have this method if you just use the more verbose version when playing?

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java [hunk 7]
```diff
@@ -68,10 +64,19 @@ public PlaylistService getService() {
...
-        this.audioPlayer = new SingleFileAudioPlayer();
+        this.audioPlayer = new SingleFileAudioPlayer(
+                new OnCompletionListener(){
+                    @Override public void onCompletion(MediaPlayer mp) {
+                        new BroadcastIntent(SingleFileAudioPlayer.ACTION_SONG_FINISHED).send(PlaylistService.this);
+                    }
+        });
+
+        this.queue = new Playlist();
+        queue.add((new MediaStoreWrapper(this)).list().get(0));
```
Comment (bryansills) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: take out the test code.

File: fanClub/src/com/lastcrusade/fanclub/util/MusicListAdapter.java (hunk 3 above)
Comment (bryansills) [inline fanClub/src/com/lastcrusade/fanclub/util/MusicListAdapter.java]: i think it was for testing. can we get back the original code?

File: fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java [hunk 8]
```diff
@@ -34,23 +49,29 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
...
-    
-                @Override
-                public void onClick(View v) {
-                    try {
-                        PlaylistService service = playlistServiceLocator.getService();
-                        if (service.isPlaying()) {
-                            service.pause();
-                        } else {
-                            service.play();
+        final ImageButton playPause = ((ImageButton) view.findViewById(R.id.btn_play_pause));
```
Comment (ejohnson44) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java]: why is the button final?

File: fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java (hunk 6 above)
Comment (ejohnson44) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java]: The method name here doesn't make much sense to me. If you are setting the image, why does it take in the whole button? Why not just the image from the button or something like that? - actually, I see what is going on. But this leads to another question- why don't you just have this take in nothing and do the whole getView() and chaining inside here instead of outside?

File: fanClub/src/com/lastcrusade/fanclub/model/Playlist.java [hunk 9]
```diff
@@ -47,4 +47,12 @@ public SongMetadata remove(int index) {
     public void setList(List<SongMetadata> list) {
         this.list = list;
     }
+    
+    public int size(){
+        return list.size();
+    }
+    
+    public SongMetadata getNext(){
+        return list.get(0);
```
Comment (ejohnson44) [inline fanClub/src/com/lastcrusade/fanclub/model/Playlist.java]: Doesn't this mean that we cannot loop through a playlist? I'm assuming that whenever a song is done playing it gets removed, but I thought we were just going to have a counter incremented or something so that we can get back to the previous songs if we have things repeating

File: fanClub/res/layout/content_frame.xml (hunk 1 above)
Comment (reidbaker) [inline fanClub/res/layout/content_frame.xml:4]: fixed

File: fanClub/src/com/lastcrusade/fanclub/CoreActivity.java (hunk 2 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/CoreActivity.java]: fixed

File: fanClub/src/com/lastcrusade/fanclub/util/MusicListAdapter.java (hunk 3 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/util/MusicListAdapter.java]: fixed

File: fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java (hunk 4 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java]: fixed

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java (hunk 5 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: It was my doing.  Do we not like name? If so what do you suggest?

File: fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java (hunk 6 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java]: fixed

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java (hunk 7 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: this was the source of my null pointer exception code :( fixed

File: fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java (hunk 8 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaybarFragment.java]: no longer final

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java (hunk 5 above)
Comment (dgreenhalgh) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: I just didn't know why it was changed. queue is likely more representative of the how it will be used, but playlist more closely aligns with the concept of a music playlist. 

Could be a design decision that gets repeated down the road. Thoughts, anyone?

Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: Changed to mPlaylist for member playlist.

File: fanClub/src/com/lastcrusade/fanclub/audio/SingleFileAudioPlayer.java [hunk 10]
```diff
@@ -32,13 +38,15 @@ public void setSongByPath(String filePath) {
...
+
     public void play() {
         try {
-            new File((new File(this.filePath).getParentFile().list())[0]).exists();
+            new File((new File(this.filePath).getParentFile().list())[0])
+                    .exists();
             if (player.isPlaying()) {
                 player.stop();
             }
+            player.reset();
```
Comment (ejohnson44) [inline fanClub/src/com/lastcrusade/fanclub/audio/SingleFileAudioPlayer.java:57]: Why are we resetting the player every time?

Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/audio/SingleFileAudioPlayer.java:57]: Because we have to or better phrased I am under the impression we have to.
You have to reset the player or there is a weird bug where the player crashes if you change the state near the end of the file.

File: fanClub/src/com/lastcrusade/fanclub/components/PlaylistFragment.java [hunk 11]
```diff
@@ -114,4 +118,32 @@ protected PlaylistService getPlaylistService() {
...
+    private class PlayListAdapter extends MusicListAdapter {
+        public PlayListAdapter(
+                Context mContext,
+                List<SongMetadata> metadataList,
+                UserList users
+                ) {
+            super(mContext, metadataList, users);
+        }
+
+        public void updateMusic(List<SongMetadata> unplayed, List<SongMetadata> played) {
```
Comment (ejohnson44) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaylistFragment.java:131]: The more I think about this, I think it makes more sense for us to just keep track of the index of the song we are currently playing (probably in playlist service?), and then just have the view reference that position and compare its current position to that position. That way we are not creating a bunch of new lists every time we hit a new song. And then perhaps songs that are not fully transferred can have some kind of flag attached to them, or they can be in some sort of parallel list.

File: fanClub/src/com/lastcrusade/fanclub/model/Playlist.java [hunk 12]
```diff
@@ -35,16 +43,56 @@ public SongMetadata remove(SongMetadata meta) {
...
+    public void setPlaylist(List<SongMetadata> unPlayed, List<SongMetadata> played){
+        this.unPlayed = unPlayed;
+        this.played = played;
+    }
+    
+    public int size(){
+        return unPlayed.size();
+    }
+    
+    public SongMetadata getHead(){
```
Comment (ejohnson44) [inline fanClub/src/com/lastcrusade/fanclub/model/Playlist.java:75]: Since you have unplayed below played, this is a little confusing - maybe change it to getNextUnplayed()?

File: fanClub/src/com/lastcrusade/fanclub/model/Playlist.java [hunk 13]
```diff
@@ -35,16 +43,56 @@ public SongMetadata remove(SongMetadata meta) {
...
+     * @return new top of unplayed list
+     */
+    public SongMetadata moveNext(){
+        if(unPlayed.size() > 0){
+            //Add top of the unPlayed list to played list
+            played.add(unPlayed.get(0));
+            //Remove the top of unplayed list
+            unPlayed.remove(0);
+        }
+        if(unPlayed.size() > 0){
```
Comment (ejohnson44) [inline fanClub/src/com/lastcrusade/fanclub/model/Playlist.java]: This is the same as the condition above... is this correct?

Actually, I think this is why your last song would never play.

Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/model/Playlist.java]: No the unplayed list shrinks at line 89. There is most likely a cleaner way to structure this but I couldnt think of it.

File: fanClub/src/com/lastcrusade/fanclub/model/Playlist.java (hunk 12 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/model/Playlist.java:75]: Yes this is left over from before there were 2 lists. I want to wait to change this until we decide on if we are keeping a playlist index. I think that might be best right now.

File: fanClub/src/com/lastcrusade/fanclub/components/PlaylistFragment.java (hunk 11 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaylistFragment.java:131]: The more I think about this the more I think you are right. Unless there are any huge objections I say after we pull this in and I will do the playlist refactor as its own branch.

Comment (ejohnson44) [inline fanClub/src/com/lastcrusade/fanclub/components/PlaylistFragment.java:131]: Sounds good to me :] I'd like to get this functionality pulled in before we work on the refactor

File: fanClub/src/com/lastcrusade/fanclub/model/Playlist.java (hunk 13 above)
Comment (ejohnson44) [inline fanClub/src/com/lastcrusade/fanclub/model/Playlist.java]: I see that it shrinks, but I still think this might be the cause of the problem since you are removing the song from unplayed before getting it again. I think that if you change it to this:

``` java
if(unPlayed.size()>0){
    SongMetadata song = unPlayed.get(0);
    played.add(song);
    unPlayed.remove(0);
    return song;
}
```

Then it will fix your problem

File: fanClub/src/com/lastcrusade/fanclub/audio/SingleFileAudioPlayer.java [hunk 14]
```diff
@@ -32,13 +38,15 @@ public void setSongByPath(String filePath) {
     public boolean isPlaying() {
         return player.isPlaying();
     }
-    
+
     public void play() {
         try {
-            new File((new File(this.filePath).getParentFile().list())[0]).exists();
+            new File((new File(this.filePath).getParentFile().list())[0])
```
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/audio/SingleFileAudioPlayer.java:52]: Why do we have to do all this?  Isn't the filePath sufficient?

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java [hunk 15]
```diff
@@ -3,37 +3,38 @@
...
+import com.lastcrusade.fanclub.util.Toaster;
+import com.lastcrusade.fanclub.R;
 
 /**
- * This service is responsible for holding the play queue, and feeding songs to
- * the Audio player service.
- * 
- * @author Jesse Rosalia
- * 
+ * This service is responsible for holding the play queue
```
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: Removed half of the javadoc comment, and it seems to still apply.

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java [hunk 16]
```diff
@@ -68,19 +66,26 @@ public PlaylistService getService() {
...
     private SingleFileAudioPlayer audioPlayer;
+    private Playlist mPlaylist;
 
     @Override
     public IBinder onBind(Intent intent) {
-        this.audioPlayer = new SingleFileAudioPlayer();
+        this.audioPlayer = new SingleFileAudioPlayer(
+                new OnCompletionListener(){
+                    @Override public void onCompletion(MediaPlayer mp) {
+                        new BroadcastIntent(SingleFileAudioPlayer.ACTION_SONG_FINISHED).send(PlaylistService.this);
```
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: It looks like we send this broadcast action, receive it several lines down, and the broadcast a different action.  Can we collapse that process and just broadcast the different action from here?

File: fanClub/res/values/error_strings.xml [hunk 17]
```diff
@@ -0,0 +1,5 @@
```
Comment (bryansills) [inline fanClub/res/values/error_strings.xml:0]: maybe rename this file errors.xml? it goes more with the naming conventions of the other xml files.

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java (hunk 15 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: Audio player service no longer exists ...

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java (hunk 16 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: I don't think we should. What is happening here is we are passing an onCompleation listener to single file audio player so that we can be informed when the song is over. When the song is over the playlist service moves to the next song and broadcasts that the playlist is updated. The playlist fragment listens to playlist updated. 

In addition to playlist service the playbar fragment listens to song finished and changes the pause play button.

Basically I think it makes sense for single file audio player to say the song is finished and the playlist service to say the playlist has changed.

File: fanClub/res/values/error_strings.xml (hunk 17 above)
Comment (reidbaker) [inline fanClub/res/values/error_strings.xml:0]: I thought about that but I wanted to say what it was expected to hold. Anyone else want to chime in?

File: fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java (hunk 15 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/service/PlaylistService.java]: Ahh good point.  But it still feeds songs to the SIngleFileAudioPlayer, so we can update the comment to reflect that.

File: fanClub/src/com/lastcrusade/fanclub/model/Playlist.java (hunk 13 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/model/Playlist.java]: The problem with not being able to play the last song was because I had `> 1` in some if not `> 0` or  `>=1`.  The problem with your code is I was intending to return the _new_ head of unPlayed not the old head. You shouldn't do that with out checking the _new_ length. 
I have looked at how the code is used however and I never take advantage of the returned value so I think I will change the return type to void.
